File: cpp/BikeDelivery.cpp
// =================================================================
// FILE: BikeDelivery.cpp
// =================================================================
#include "BikeDelivery.h"
#include "Configuration.h"
#include "Checker.h"
#include <iostream>
#include <stdexcept>

// Calculates the Euclidean distance between two points.
double BikeDelivery::calculateDistance(const Point& source, const Point& destination) {
    double dx = source.x - destination.x;
    double dy = source.y - destination.y;
    return std::sqrt(dx * dx + dy * dy);
}

// Implements price calculation for BikeDelivery.
int BikeDelivery::calculatePrice(const Parcel& parcel, bool fastDelivery) {
    Configuration& config = Configuration::getInstance();
    double distance = calculateDistance(parcel.getSource(), parcel.getDestination());

    // Base price calculation
    double price = (config.getUnitPrice() * distance) 
                 + config.getBasePrice() 
                 + (0.10 * parcel.getInsurancePrice());

    // Add surcharge for fast delivery
    if (fastDelivery) {
        price += 0.10 * (config.getUnitPrice() * distance);
    }

    return static_cast<int>(price);
}

// Implements the base version of calculatePrice, defaulting to standard delivery.
int BikeDelivery::calculatePrice(const Parcel& parcel) {
    return calculatePrice(parcel, false);
}

// Implements condition checking using a specific chain of checkers for bike delivery.
void BikeDelivery::checkConditions(const Parcel& parcel) {
    std::cout << "Checking parcel conditions for Bike Delivery..." << std::endl;

    // For BikeDelivery, the check order is defined here:
    // 1. Not flammable -> 2. Not chemical -> 3. Weight -> 4. Dimensions
    FlammableChecker flammableChecker;
    ChemicalChecker chemicalChecker;
    WeightChecker weightChecker;
    DimensionChecker dimensionChecker;

    // Build the chain of responsibility.
    flammableChecker.setNextChecker(&chemicalChecker);
    chemicalChecker.setNextChecker(&weightChecker);
    weightChecker.setNextChecker(&dimensionChecker);

    try {
        // Start the validation from the first checker in the chain.
        flammableChecker.check(parcel);
        std::cout << "All conditions met. Parcel is valid for bike delivery." << std::endl;
    } catch (const std::runtime_error& e) {
        // The process stops here. The error message is already printed by the specific checker.
        // We catch the exception to prevent the program from crashing.
    }
}

File: python/bike_delivery.py
import math
from point import Point
from parcel import Parcel
from framework import UrbanDeliveryService
from checkers import FlammableChecker, DimensionChecker, WeightChecker
from configuration import Configuration

class BikeDelivery(UrbanDeliveryService):
    def _calculate_distance(self, source: Point, destination: Point) -> float:
        return math.sqrt((source.x - destination.x)**2 + (source.y - destination.y)**2)

    def check_conditions(self, parcel: Parcel) -> None:
        """Builds the chain of checkers and initiates the validation."""
        print("Checking parcel conditions for Bike Delivery...")
        
        flammable_checker = FlammableChecker()
        dimension_checker = DimensionChecker()
        weight_checker = WeightChecker()
        
        flammable_checker.set_next_checker(dimension_checker).set_next_checker(weight_checker)
        
        try:
            flammable_checker.check(parcel)
            print("All conditions met. Parcel is valid for bike delivery.")
        except ValueError:
            pass

    def calculate_price(self, parcel: Parcel, fast_delivery: bool = False) -> int:
        """Calculates price for bike delivery, with an option for fast delivery."""
        config = Configuration.get_instance()
        distance = self._calculate_distance(parcel.source, parcel.destination)
        price = (config.unit_price * distance) + config.base_price + (0.10 * parcel.insurance_price)
        if fast_delivery:
            price += 0.10 * (config.unit_price * distance)
        return int(price)